Feature/initrd

DrDeano opened the pull request:
Closes #201

SamTebbs33 commented on build.zig:
```diff
@@ -59,11 +63,28 @@ pub fn build(b: *Builder) !void {
     };
     make_iso.step.dependOn(&exec.step);
 
+    const int_size: u16 = switch (target.getCpuArch()) {
+        .i386 => 32,
+        else => unreachable,
+    };
```
Can this be replaced by `@sizeOf(usize)`?

SamTebbs33 replied:
Scratch that. Obviously that would use the host system's `usize`!

DrDeano replied:
I wanted this, but the build file is ran as 64bit so `@sizeOf(usize)` return 8 instead of 4

DrDeano replied:
I could pass the target to the ramdisk step to make this easier, plus can use the native endianness

SamTebbs33 commented on build.zig:
```diff
@@ -149,3 +170,110 @@ pub fn build(b: *Builder) !void {
+        var ramdisk = try fs.cwd().createFile(self.out_file_path, .{});
+        defer ramdisk.close();
+
+        // First write the number of files/headers
+        switch (self.int_size) {
+            32 => {
+                std.debug.assert(self.files.len < std.math.maxInt(u32));
+                try ramdisk.writer().writeIntNative(u32, @truncate(u32, self.files.len));
+            },
+            else => unreachable,
+        }
```
Perhaps passing the integer type to the RuntimeStep rather than the integer size would be nice? That would remove the need for all the switch statements.

DrDeano replied:
I did try this, but need comptime arch which isn't provided, instead will pass the target to the ramdisk step

SamTebbs33 commented on build.zig:
```diff
@@ -149,3 +170,110 @@ pub fn build(b: *Builder) !void {
+        switch (self.int_size) {
+            32 => {
+                std.debug.assert(self.files.len < std.math.maxInt(u32));
+                try ramdisk.writer().writeIntNative(u32, @truncate(u32, self.files.len));
```
Is it possible to write the integer in the target's endianness rather than the native endianness?

DrDeano replied:
Will pass the target to the ramdisk step

SamTebbs33 commented on src/kernel/initrd.zig:
```diff
@@ -0,0 +1,673 @@
+            .NO_CREATION => {
+                for (self.files) |*file| {
+                    if (std.mem.eql(u8, file.name, name)) {
+                        // Opening 2 files of the same name, will create 2 different Nodes
```
To avoid creating two nodes for the same file, you could create a list of nodes when initialising the ramdisk and then fetch from that when opening a node. That would also make destroying it upon closing unnecessary.

DrDeano replied:
I wanted to create the *Nodes on the fly so not, this would save memory but would be slower. Having all the *Nodes created at init would increase memory, but would be faster. I don't mind having 2 different nodes for the same file. I can iterate through the hash map and search for any node that is opened (I did have this) but this would make opening even slower. What are your thoughts on this?

SamTebbs33 replied:
To be honest, at the moment I can't think of any downsides to have two nodes per file (other than the extra memory allocated) so I'm gonna leave it up to you :)

DrDeano replied:
I guess as the initrd will only be used once and load all files in one go, I could create all the Nodes and add a `free` or `deinit` to clear the initrd and reuse the memory after everything is loaded. This will also free the raw ramdisk so have all the memories.

SamTebbs33 replied:
That sounds good :)

SamTebbs33 commented on src/kernel/initrd.zig:
```diff
@@ -0,0 +1,673 @@
+    fn write(fs: *const vfs.FileSystem, node: *const vfs.FileNode, bytes: []const u8) (Allocator.Error || vfs.Error)!void {}
+
+    /// See vfs.FileSystem.open
+    fn open(fs: *const vfs.FileSystem, dir: *const vfs.DirNode, name: []const u8, flags: vfs.OpenFlags) (Allocator.Error || vfs.Error)!*vfs.Node {
```
It doesn't look like this uses the `dir` argument. Is it intended that the ramdisk doesn't support directories? That's fine, I'm just asking to be up to date.

DrDeano replied:
It isn't used no, I wasn't sure how to use it anyway as yes, there are no directories and from your example, it is used for searching, but if there is a use for it, I would love to hear

SamTebbs33 replied:
Well the directory node is the directory under which the file (`name` is a file name and extension, not a path) is created/opened, so it can be ignored if you don't plan for the ramdisk to support directories at the moment.

DrDeano replied:
I see, the initrd will probably not have any directories, just to make it easier. In that case, shall I rename RamdiskFS to InitrdFS.

SamTebbs33 replied:
I'm cool with either

SamTebbs33 commented on src/kernel/initrd.zig:
```diff
@@ -0,0 +1,673 @@
+    ///     error.InvalidRamDisk - If the provided raw ramdisk is invalid. This can be due to a
+    ///                            mis-match of the number of files to the length of the raw
+    ///                            ramdisk or the wrong length provided to cause undefined parsed
+    ///                            lengths for other pars of the ramdisk.
```
s/pars/parts

SamTebbs33 commented on src/kernel/initrd.zig:
```diff
@@ -0,0 +1,673 @@
+        const ramdisk_bytes = @intToPtr([*]u8, rd_module.region.start)[0..rd_len];
+        // First @sizeOf(usize) bytes is the number of files
+        const num_of_files = std.mem.readIntSlice(usize, ramdisk_bytes[0..], builtin.endian);
+        var headers = try allocator.alloc(RamdiskHeader, num_of_files);
```
Modifying the ramdisk structure to start with the number of headers, then all of the headers themselves and after that the file content would put the headers in a packed array meaning you can just cast the ramdisk start address (plus the size of the field for the number of headers) to a `[*]RamdiskHeader` and make a slice of that, instead of allocating memory and copying everything over (it's already in memory so let's take advantage of that). What do you think?

DrDeano replied:
That is an idea. But we would need to know the name and content length to create the []u8 string. So just casting wouldn't work.

SamTebbs33 commented on src/kernel/vfs.zig:
```diff
@@ -383,6 +387,16 @@ pub fn isAbsolute(path: []const u8) bool {
+/// Arguments:
+///     IN node: *Node - The node to initialise the root node.
+///
+pub fn init(node: *Node) void {
```
Since we foresee this being used to switch root from the initrd to a disk filesystem at runtime, perhaps it should be called `setRoot` or similar?

SamTebbs33 reviewed (changes requested):
Looking great :) I was thinking about the name, and what do you think about the file extension being "ramdisk" or "rd" since not all ramdisks will be the "initial" ramdisk (which initrd is short for?)?

DrDeano committed:
Created ramdisk build step


Refactored tests for the scheduler and task


Revert "Refactored tests for the scheduler and task"

This reverts commit 2bf56a368bc18f2bd2d33c385e3672d07e4431d9.

Refactored tests for the scheduler and task


Task fmt


Task fmt again >:(


Ramdisk


Added NotOpened error for file read and write
Added vfs init to initialise the root node
Added the ramdisk.initrd file to grub
Update makeiso to copy the ramdisk to the modules folder
Add a ramdisk step to create a ramdisk to be leaded by grub and parsed by the kernel
Add test files for runtime tests of ramdisk

vfs.init => vfs.setRoot


Improved ramdisk step

Also spelling

Changed name for the initrd


Rename RamdiskFS => InitrdFS


Add deinit for initrd

Fixed VMM unmap

{}
